### PR TITLE
Add unresolved elemements plus metadata properties

### PR DESCRIPTION
**Pull request by lonnybastien**
Refactor to support unresolved elements. This requires changing all calls from GetReferenced to GetReferencedRelationshipInstances for Peer and Hierarchical relationships. With methods now returning unresolved elements, a new Reference sub class for each Model type is defined. The reference sub types adapts a ModelRelationshipInstance instead of a TSQLObject. Adapting a ModelRelationshipInstance instead of a TSQLObject allows the name of referenced elements to be exposed even if the element is unresolved. It also allows exposing the metadata on relationships (Such as Ascending on the Index.Columns relationship).

For each relationship that has metadata, a new sub type is defined with the metadata properties. This results in the Index.Columns relationship returning an IEnumerable<IndexColumnsReference> where IndexColumnsReference derives from TSQLColumnReference which in-turn derives from TSQLColum.
Metadata properties on TSQLObjecs are now exposed. One example of this is TSQLColumn now exposes a ColumnType property which defines if it is a Column, ComputedColumn or ColumnSet.
NOTE: In order to facilitate unresolved references, a few edge cases arose. For relationships that returned marker interfaces like ISqlObjectAuthorizer, there was no concrete type known at generation time to instantiate. To work around this issue, a new optional attribute was added to the modelmetadata.xml: UnresolvedReturnType. The UnresolvedReturnType attribute defines the type to instantiate when an unresolved element is found so the return type will still implement the type defined in the ReturnType attribute. This pattern is a little unconventional, however, to enforce the strongly typed returns values of relationships these “Unresolved” types needed to be created to bridge the gap between the strongly typed API and the unresolved nature of the model.

### CODE REVIEW

File: DacFxStronglyTypedModel/Utils.tt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿//------------------------------------------------------------------------------
+﻿<#//------------------------------------------------------------------------------
```
Comment (kevcunnane) [inline DacFxStronglyTypedModel/Utils.tt]: Why are you removing the copyright?

File: DacFxStronglyTypedModel/Interfaces.tt [hunk 2]
```diff
@@ -57,6 +57,9 @@ foreach(SqlServerVersion currentSQLVersion in GetSqlServerVersionValues())
 		overrides.TryGetValue(type.Name, out modelEntry);
 		
 #>
+    public interface <#= className #>Reference : <#= className #>
+	{
```
Comment (kevcunnane) [inline DacFxStronglyTypedModel/Interfaces.tt]: Formatting seems off?

File: DacFxStronglyTypedModel/Utils.tt [hunk 3]
```diff
@@ -23,7 +23,7 @@
 //    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 //    SOFTWARE.
 //</copyright>
-//------------------------------------------------------------------------------
+//------------------------------------------------------------------------------#>
```
Comment (kevcunnane) [inline DacFxStronglyTypedModel/Utils.tt]: Why is this added?

File: DacFxStronglyTypedModel/Utils.tt [hunk 4]
```diff
@@ -32,61 +32,36 @@
 <#@ import namespace="System.Xml" #>
 <#@ import namespace="System.Collections.Generic" #>
 <#@ import namespace="Microsoft.SqlServer.Dac.Model"#>
+<#@ import namespace="System.Text.RegularExpressions"#>
 <#+
-//------------------------------------------------------------------------------
```
Comment (kevcunnane) [inline DacFxStronglyTypedModel/Utils.tt]: Again is there a reason we're removing copyright from the generated files?

File: DacFxStronglyTypedModel/Utils.tt [hunk 5]
```diff
@@ -125,9 +100,12 @@ public class RelationshipOverride
 
 	public string RelationshipName {get;set;}
 	public string ReturnType {get;set;}
-	public string ReturnTypeNamespace {get;set;}
-	public bool Specialize {get;set;}
-	public bool AdaptInstance {get;set;}
+    public string InternalReturnType {get; set;}
+    public string InternalReturnTypeNamespace {get; set;}
+	public string ReturnTypeNamespace {get; set;}
```
Comment (kevcunnane) [inline DacFxStronglyTypedModel/Utils.tt]: Is formatting wrong?

File: DacFxStronglyTypedModel/Utils.tt (hunk 1 above)
Comment (lonnybastien) [inline DacFxStronglyTypedModel/Utils.tt]: This was done because the utils.tt output file, util.cs, outputs no content and has no need for the copyright to be outputted. Because utils.tt is included in the other template files which already output the copyright, it was outputting the copyright twice in those output files. Adding <# keeps the copyright at the top of util.tt file and removes the duplication from the other files.

File: DacFxStronglyTypedModel/Utils.tt (hunk 3 above)
Comment (lonnybastien) [inline DacFxStronglyTypedModel/Utils.tt]: See the comment above where I added the "<#". This is just the closing tag.
